plugins/ssr/ssrelim.ml

```diff
@@ -253,31 +253,48 @@ let ssrelim ?(is_case=false) deps what ?elim eqid elim_intro_tac =
   (* Here we try to understand if the main pattern/term the user gave is
    * the first pattern to be matched (i.e. if elimty ends in P t1 .. tn,
    * weather tn is the t the user wrote in 'elim: t' *)
-  let c_is_head_p, gl = match cty with
+  let c_is_head_p, gl =
+    match cty with
     | None -> true, gl  (* The user wrote elim: _ *)
     | Some (c, c_ty, _) ->
-    let res = 
-      (* we try to see if c unifies with the last arg of elim *)
-      if elim_is_dep then None else
-      let arg = List.assoc (n_elim_args - 1) elim_args in
-      let gl, arg_ty = pfe_type_of gl arg in
-      match saturate_until gl c c_ty (fun c c_ty gl ->
-        pf_unify_HO (pf_unify_HO gl c_ty arg_ty) arg c) with
-      | Some (c, _, _, gl) -> Some (false, gl)
-      | None -> None in
-    match res with
-    | Some x -> x
-    | None ->
-      (* we try to see if c unifies with the last inferred pattern *)
-      let inf_arg = List.hd inf_deps_r in
-      let gl, inf_arg_ty = pfe_type_of gl inf_arg in
-      match saturate_until gl c c_ty (fun _ c_ty gl ->
-              pf_unify_HO gl c_ty inf_arg_ty) with
-      | Some (c, _, _,gl) -> true, gl
-      | None ->
-        errorstrm Pp.(str"Unable to apply the eliminator to the term"++
-          spc()++pr_econstr_env env (project gl) c++spc()++str"or to unify it's type with"++
-          pr_econstr_env env (project gl) inf_arg_ty) in
+      let rec first = function
+        | [] ->
+          errorstrm Pp.(str"Unable to apply the eliminator to the term"++
+            spc()++pr_econstr_env env (project gl) c++spc())
+        | x :: rest ->
+          match x () with
+          | None -> first rest
+          | Some (b,gl) -> b, gl
+      in
+      (* unify two terms if their heads are not applied unif variables, eg
+       * a HO unif problem *)
+      let pf_unify_HO_rigid gl a b =
+        let is_HO_unif x = match EConstr.kind (project gl) x with
```

Comment:
`is_HO_unif` is a bit confusing, could we find a better name?

Reply:
Should be better now.